common/flatpak-instance.c

```diff
@@ -429,6 +435,354 @@ flatpak_instance_new_for_id (const char *id)
+   * per-app directory, rather than being a symlink. Open it with
+   * O_NOFOLLOW to make sure it is indeed not a symlink. */
+  tmp = g_build_filename (per_app_dir, "tmp", NULL);
+  tmp_fd = openat (AT_FDCWD, tmp, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+
+  if (tmp_fd >= 0)
+    {
+      glnx_autofd int lock_fd = openat (tmp_fd, ".flatpak-tmpdir",
+                                        O_RDONLY | O_CLOEXEC);
+
+      if (lock_fd >= 0 && fcntl (lock_fd, F_SETLK, &exclusive_lock) == 0)
+        glnx_shutil_rm_rf_at (AT_FDCWD, tmp, NULL, NULL);
+    }
+
+  for (i = 0; i < G_N_ELEMENTS (temp_directories); i++)
+    {
+      g_autofree char *link_path = g_build_filename (per_app_dir,
+                                                     temp_directories[i].link_name,
+                                                     NULL);
+      g_autofree char *path = NULL;
+      const char *purpose = temp_directories[i].purpose;
+      glnx_autofd int lock_fd = -1;
+
+      if (flatpak_instance_reuse_per_app_tmp_directory (app_id,
+                                                        link_path,
+                                                        purpose,
+                                                        F_WRLCK,
+                                                        &lock_fd,
+                                                        &path))
+        {
+          g_assert (g_str_has_prefix (path, purpose));
+          glnx_shutil_rm_rf_at (AT_FDCWD, path, NULL, NULL);
+          unlink (link_path);
+        }
+    }
+}
+
+gboolean
+flatpak_instance_setup_per_app_tmp_directory (const char *app_id,
+                                              const char *purpose,
+                                              int *lock_fd_out,
+                                              char **path_out,
+                                              GError **error)
+{
+  g_autofree char *link_path = NULL;
+  g_autofree char *lock = NULL;
+  g_autofree char *path = NULL;
+  g_autofree char *per_app_dir = NULL;
+  g_autofree char *per_app_dir_lock = NULL;
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  glnx_autofd int lock_fd = -1;
+  glnx_autofd int per_app_dir_lock_fd = -1;
+  const char *link_name;
+  struct flock exclusive_lock = {
+      .l_type = F_WRLCK,
+      .l_whence = SEEK_SET,
+      .l_start = 0,
+      .l_len = 0,
+  };
+  struct flock shared_lock = {
+      .l_type = F_RDLCK,
+      .l_whence = SEEK_SET,
+      .l_start = 0,
+      .l_len = 0,
+  };
+
+  if (strcmp (purpose, "/dev/shm") == 0)
+    link_name = "dev-shm";
+  else if (strcmp (purpose, "/tmp") == 0)
+    link_name = "tmp";
+  else
+    g_return_val_if_reached (FALSE);
+
+  /* These app IDs are in the same directory as per-instance directories,
+   * but they can't collide, because per-instance directories are integers
+   * with no dots, whereas app IDs have at least two dots */
```

Comment:
I could use a different directory if you prefer. Mixing app IDs with instance IDs, and discriminating by whether it's an integer (=> instance ID) or not (=> app ID), is a bit ugly.

I could put the app's `/tmp` in `$XDG_RUNTIME_DIR/app/$FLATPAK_ID/tmp` instead of in `$XDG_RUNTIME_DIR/.flatpak/$FLATPAK_ID/tmp`, if you prefer.

However, I specifically *don't* want to move the symlink pointing to the app's private `/dev/shm` (really `/dev/shm/flatpak-$FLATPAK_ID-XXXXXX`) from `$XDG_RUNTIME_DIR/.flatpak/$FLATPAK_ID/dev-shm` to `$XDG_RUNTIME_DIR/app/$FLATPAK_ID/dev-shm`, because if I did, the app could alter where it points, causing badness (probably just denial of service, but in the worst case, privilege escalation) on the host system - so that one should stay where it is.

If `/tmp` becomes `/tmp/flatpak-$FLATPAK_ID-XXXXXX`, then the symlink pointing to that should also be somewhere outside the app's control.